wled00/wled_server.cpp

```diff
@@ -404,59 +400,47 @@ void initServer()
   });
 
   server.on(_update, HTTP_POST, [](AsyncWebServerRequest *request){
-    if (!correctPIN) {
-      serveSettings(request, true); // handle PIN page POST request
-      return;
-    }
-    if (otaLock) {
-      serveMessage(request, 401, FPSTR(s_accessdenied), FPSTR(s_unlock_ota), 254);
-      return;
-    }
-    if (Update.hasError()) {
-      serveMessage(request, 500, F("Update failed!"), F("Please check your file and retry!"), 254);
+    if (request->_tempObject) {
+      auto ota_result = getOTAResult(request);
+      if (ota_result.first) {
+        if (ota_result.second.length() > 0) {
+          serveMessage(request, 500, F("Update failed!"), ota_result.second, 254);
```

Comment:
@willmmiles another idea to help users understand why OTA has failed is to display error strings from ``Update.errorString()`` in the web UI. I have made some experiments with this in WLED-MM, and the error strings are really useful to understand common failure causes like "wrong CPU", "OTA partition too small", "could not activate firmware" etc.

https://github.com/MoonModules/WLED-MM/blob/mdev/wled00/wled_server.cpp#L332-L345